packages/fast-figma-plugin-msft/src/core/node.ts

```diff
@@ -0,0 +1,128 @@
+ * for each design tool.
+ */
+export abstract class PluginNode {
+    public abstract id: string;
+    public abstract type: string;
+    public abstract children(): PluginNode[];
+    public abstract parent(): PluginNode | null;
+    public abstract supports(): Array<RecipeTypes | "designSystem">;
+
+    /**
+     * Retrieves the contextual design system for the node
+     */
+    public get designSystem(): Partial<DesignSystem> {
+        // Retrun value from the cache if we have it
+        if (DesignSystemCache.has(this.id)) {
+            return DesignSystemCache.get(this.id) as Partial<DesignSystem>;
+        }
+
+        let node: PluginNode | null = this;
+        const designSystems: Array<Partial<DesignSystem>> = [];
+
+        while (node !== null) {
+            designSystems.push(node.getPluginData("designSystem"));
+            node = node.parent();
+        }
+
+        const designSystem = designSystems.reduceRight((prev, next) => ({
+            ...prev,
+            ...next,
+        }));
+        DesignSystemCache.set(this.id, designSystem);
+
+        return designSystem;
+    }
+
+    public get recipes(): string[] {
+        const recipes = this.getPluginData("recipes");
+        return recipes;
+    }
+
+    public set recipes(recipes: string[]) {
+        this.setPluginData("recipes", recipes);
+    }
+
+    /**
+     * Set a property of the design system on this node
+     * @param key - the design system property name
+     * @param value - the design system property value
+     */
+    public setDesignSystemProperty<K extends keyof DesignSystem>(
+        key: K,
+        value: DesignSystem[K]
+    ): void {
+        this.setPluginData("designSystem", {
+            ...this.getPluginData("designSystem"),
+            [key]: value,
+        });
+
+        this.invalidateDesignSystemCache();
+    }
+
+    public deleteDesignSystemProperty<K extends keyof DesignSystem>(key: K): void {
+        const data = this.getPluginData("designSystem");
+        delete data[key];
+        this.setPluginData("designSystem", data);
+        this.invalidateDesignSystemCache();
+    }
+
+    /**
+     * Returns all design system overrides applied to the node
+     */
+    public get designSystemOverrides(): Partial<DesignSystem> {
+        return this.getPluginData("designSystem");
+    }
+
+    public abstract paint(data: RecipeData): void;
+    protected abstract getPluginData<K extends keyof PluginNodeData>(
+        key: K
+    ): PluginNodeData[K];
+    protected abstract setPluginData<K extends keyof PluginNodeData>(
+        key: K,
+        value: PluginNodeData[K]
+    ): void;
+
+    /**
+     * Retrieve the effective background color for the node.
+     * This color is communicated to color recipes as the
+     * backgroundColor context for a node
+     */
+    public abstract getEffectiveBackgroundColor(): ColorRGBA64;
+
+    public invalidateDesignSystemCache() {

```

Comment:
can we get a code comment on this to let us know what happens when the cache is invalidated?